exercises/bracket-push/example.rs

```diff
@@ -0,0 +1,84 @@
+use std::collections::HashMap;
+
+pub struct Brackets {
+    raw_brackets: Vec<char>,
+    pairs: MatchingBrackets,
+}
+
+impl<'a> From<&'a str> for Brackets {
+    fn from(i: &str) -> Self {
+        Brackets::new(String::from(i), None)
+    }
+}
+
+pub struct MatchingBrackets {
+    collection: HashMap<char, char>,
+}
+
+impl From<Vec<(char, char)>> for MatchingBrackets {
+    fn from(i: Vec<(char, char)>) -> Self {
+        let mut c: HashMap<char, char> = HashMap::new();
+
+        for (k, v) in i {
+            c.insert(k, v);
+        }
+
+        MatchingBrackets { collection: c }
+    }
+}
+
+impl MatchingBrackets {
+    fn contains(&self, other: &char) -> bool {
+        let known = self.collection.keys().chain(self.collection.values()).collect::<Vec<_>>();
+        known.contains(&other)
+    }
+
+    fn closer_for(&self, k: &char) -> Option<&char> {
+        self.collection.get(k)
+    }
+
+    fn unmatched(&self, open: char, potential_close: char) -> Vec<char> {
+        let mut ret: Vec<char> = Vec::new();
+
+        if let Some(closing_bracket) = self.closer_for(&open) {
+            if *closing_bracket != potential_close {
+                ret.push(open);
+                ret.push(potential_close);
+            }
+        } else {
+            ret.push(open);
+            ret.push(potential_close);
+        }
+        ret
+    }
+}
+
+impl Brackets {
+    pub fn new(s: String, pairs: Option<Vec<(char, char)>>) -> Self {
+        let p = match pairs {
+            Some(x) => MatchingBrackets::from(x),
+            None => MatchingBrackets::from(vec![('[', ']'), ('{', '}'), ('(', ')')]),
+        };
+
+        Brackets {
+            raw_brackets: s.chars().filter(|c| p.contains(&c)).collect::<Vec<char>>(),
+            pairs: p,
+        }
+    }
+
+    pub fn are_balanced(&self) -> bool {
+        let mut unclosed: Vec<char> = Vec::new();
+
+        for &bracket in self.raw_brackets.iter() {
+            if let Some(last_unclosed) = unclosed.pop() {
+                for c in self.pairs.unmatched(last_unclosed, bracket) {
```

Comment:
take a look at whether this can be expressed as an [`extend`](https://doc.rust-lang.org/std/iter/trait.Extend.html) which I think cn work because the docs say `impl<T> Extend<T> for Vec<T>`.

the idea of letting the `pairs`  determine what's unmatched seems OK.